packages/lit-dev-server/src/middleware/fake-github-middleware.ts

```diff
@@ -0,0 +1,133 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import type Koa from 'koa';
+
+/**
+ * Options for fakeGitHubMiddleware.
+ */
+export interface FakeGitHubMiddlewareOptions {
+  /**
+   * GitHub OAuth App client ID.
+   *
+   * In reality, this is generated when a GitHub OAuth App is first created. For
+   * this fake, it can be any string.
+   */
+  clientId: string;
+
+  /**
+   * GitHub OAuth app secret.
+   *
+   * In reality, this is generated from the GitHub OAuth App settings screen.
+   * For this fake, it can be any string.
+   */
+  clientSecret: string;
+
+  /**
+   * URL to redirect to after app authentication or installation.
+   *
+   * In reality, this is configured in the GitHub OAuth App settings screen. For
+   * this fake, it can be any URL.
+   */
+  redirectUrl: string;
+}
+
+/**
+ * Koa middleware that simulates various GitHub APIs and authentication screens
+ * used by the lit.dev Playground.
+ */
+export const fakeGitHubMiddleware = (
+  options: FakeGitHubMiddlewareOptions
+): Koa.Middleware => {
+  let fake = new FakeGitHub(options);
+  return async (ctx, next) => {
+    if (ctx.path === '/reset') {
+      return fake.reset(ctx);
+    } else if (ctx.path === '/login/oauth/authorize') {
+      return fake.authorize(ctx);
+    } else {
+      return next();
+    }
+  };
+};
+
+const randomString = () => String(Math.floor(Math.random() * 1e10));
+
+class FakeGitHub {
+  private readonly _options: FakeGitHubMiddlewareOptions;
+  private readonly _userToCode = new Map<string, string>();
+
+  constructor(options: FakeGitHubMiddlewareOptions) {
+    this._options = options;
+  }
+
+  /**
+   * Generate a random user ID for this session and persist it in a cookie.
+   */
+  private _getOrSetUserId(ctx: Koa.Context) {
+    let userId = ctx.cookies.get('userid');
+    if (!userId) {
+      userId = randomString();
+      ctx.cookies.set('userid', userId);
+    }
+    return userId;
+  }
+
+  /**
+   * Clear all server state and browser cookies.
+   */
+  reset(ctx: Koa.Context) {
+    this._userToCode.clear();
+    ctx.cookies.set('userid', null);
+    ctx.cookies.set('authorized', null);
+    ctx.status = 200;
+    ctx.body = 'fake github successfully reset';
+  }
+
+  /**
+   * Simulates https://github.com/login/oauth/authorize, the screen that prompts
+   * a user to authorize a GitHub OAuth App.
+   *
+   * Documentation:
+   * https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#1-request-a-users-github-identity
+   */
+  authorize(ctx: Koa.Context) {
```

Comment:
Since this is a fake, is there any benefit to add extra verification/assertions such as ensuring that the `ctx` has a `scope`/`state` set for example?
Testing against a more constrained endpoint to restrict what we do.
non actionable comment.

Reply:
Seems reasonable. Added a check for `scope=gist`. Nothing for `state` for now, see other comment.